service/dap/server_test.go

```diff
@@ -1669,6 +1669,87 @@ func TestSetBreakpoint(t *testing.T) {
 	})
 }
 
+// TestLaunchSubstitutePath sets a breakpoint using a path
+// that does not exist and expects the substitutePath attribute
+// in the launch configuration to take care of the mapping.
+func TestLaunchSubstitutePath(t *testing.T) {
+	runTest(t, "loopprog", func(client *daptest.Client, fixture protest.Fixture) {
+		localDirPath := filepath.Join(string(filepath.Separator), "path", "that", "does", "not", "exist")
+		if runtime.GOOS == "windows" {
+			localDirPath = "C:" + localDirPath
+		}
+		localFilePath := filepath.Join(localDirPath, "loopprog.go")
+		debuggerDirPath := filepath.Dir(fixture.Source)
```

Comment:
To improve readability, I would drop this var and rename the other one:
`"from": nonExistentDir, "to": filepath.Dir(fixture.Source)`
Also, I am confused why you have one as file path and another as dir path. If it doesn't exist, that won't matter either way, but what if it did? When does file vs dir matter?

Reply:
The path mappings have to be directories, but we use the file path to set the breakpoint.